### PR TITLE
Implementing mypyc support pt. 2 

### PR DESCRIPTION
**Pull request by ichard26**
It's finally happening :tada:, the initial work was done several years ago in GH-1009 by @ msullivan is now being followed through by yours truly. If this work doesn't find its way to PyPI eventually, I'll be disappointed :slightly_smiling_face: 

### Performance

I wrote an entire report which can be read here: https://gist.github.com/ichard26/b996ccf410422b44fcd80fb158e05b0d, but the TL;DR is:

- Formatting _with_ safety checks: **1.93x faster**
- Formatting _without_ safety checks: **2.11x faster**
- Blib2to3 parsing: 1.81x faster
- Import (import black): **1.16x faster**

Windows numbers are to come later. Unfortunately I probably won't be able to get MacOS numbers (I own no MacOS running machine), but I'll see what I can do.

### Stability

I've been running the test suite during the wheel build and as of latest of this branch it's all green. Though it should be noted that the arm64 and the arm portion of the universial2 wheels are completely untested since we don't have access to any M1 CI resources :(

Locally I've ran [diff-shades](https://github.com/ichard26/black-mypyc-wheels/blob/main/diff_shades.py) (Jelle, it's basically mypy-primer but for black) against the CPython 3.8 manylinux wheel and no difference between interpreted and compiled Black was found (across the same project suite used by black-primer + blackbench :wink:). I plan to also test Windows with diff-shades in the following days. I'll also get Cooper to test Intel MacOS + M1 MacOS for me, but I need to come up with better instructions before that.

Anyhow, please don't expect this to be fully tested. I plan to ask the wider community to test the experimental compiled wheels (and push any additional fixes) after this PR is merged. 

### Review notes

This PR contains a lot of things including QOL improvements, optimizations, compatibility changes, hacks, etc. While the diff isn't that painful, following the commit history should give you some context on why certain changes were made.

It should be noted that I plan to handle the merge of this PR myself since this project involves a lot of moving parts and I'd like to maintain control for the time being. Thanks!

---

Related issue: GH-366  
Related repo: https://github.com/ichard26/black-mypyc-wheels

And if you're some passerby who really wants to try the experimental wheels, checkout the workflows here: https://github.com/ichard26/black-mypyc-wheels/actions.

### CODE REVIEW

**Comment by cooperlees** [comment]
Nice work. FWIW - I'm totally down for `black-primer` to be refactored to do what it does today as well as what diff-shades does to cut down on the number of tools we all use everywhere. I guess diff-shades like behavior could be more easily added to our CI too if we incorporated it into the black repo.

File: mypy.ini [hunk 1]
```diff
@@ -33,7 +36,23 @@ cache_dir=/dev/null
...
+# TOML people like bare CR as a newline sequence I suppose). Anyway,
+# some versions of tomli still specify str as the type so mypy complains
+# that the type: ignore is unnecessary. Gosh I wish I wasn't telling
+# this story.
+#
+# See also: https://github.com/psf/black/pull/2408
+#           https://github.com/pypa/pip/pull/10238 (because black is
+#             following whatever they do)
+#           https://github.com/toml-lang/toml/issues/837
+warn_unused_ignores=False
```
Comment (hukkin) [inline mypy.ini:58]: Now this is an exciting PR, good job!

I sense some frustration towards this particular ignore :smile: 
If you want you can remove all of this by going to `black.files.py` and replacing
```python
with open(path_config, encoding="utf8") as f:
    pyproject_toml = tomli.load(f)  # type: ignore  # due to deprecated API usage
```
with
```python
with open(path_config, encoding="utf8") as f:
    pyproject_toml = tomli.loads(f.read())
```

The behavior 100% the same, no need to worry about that.

File: src/black/brackets.py [hunk 2]
```diff
@@ -49,7 +49,9 @@
 DOT_PRIORITY: Final = 1
 
 
-class BracketMatchError(KeyError):
+# Ideally this would be a subclass of KeyError, but mypyc doesn't like that.
```
Comment (JelleZijlstra) [inline src/black/brackets.py:52]: I don't think it really needs to be a subclass of KeyError. I say we just remove this comment and leave it a direct subclass of Exception.

File: src/black/linegen.py [hunk 3]
```diff
@@ -362,6 +365,11 @@ def rhs(line: Line, features: Collection[Feature]) -> Iterator[Line]:
                 line, line_length=mode.line_length, features=features
             )
 
+        # HACK: functions (like rhs) compiled by mypyc don't retain their __name__
```
Comment (JelleZijlstra) [inline src/black/linegen.py:368]: I guess it would be a bigger refactor but I'd like it better if we just stopped using `__name__`.

File: src/black/parsing.py [hunk 4]
```diff
@@ -79,7 +91,7 @@ def lib2to3_parse(src_txt: str, target_versions: Iterable[TargetVersion] = ()) -
         src_txt += "\n"
 
     for grammar in get_grammars(set(target_versions)):
-        drv = driver.Driver(grammar, pytree.convert)
+        drv = driver.Driver(grammar)
```
Comment (JelleZijlstra) [inline src/black/parsing.py:94]: What is the motivation behind this change?

File: src/black/strings.py [hunk 5]
```diff
@@ -142,6 +152,11 @@ def normalize_string_prefix(s: str, remove_u_prefix: bool = False) -> str:
     return f"{new_prefix}{match.group(2)}"
 
 
+@lru_cache(maxsize=256)
```
Comment (JelleZijlstra) [inline src/black/strings.py:155]: Does this really help? `re.compile` already caches things internally I believe.

**Review by JelleZijlstra — APPROVED** [review]
Thanks for all your work on this! I'm really excited to see this PR.

File: src/black/parsing.py (hunk 4 above)
Comment (JelleZijlstra) [reply src/black/parsing.py:94]: Never mind it's explained in blib2to3 code.

File: src/black/brackets.py (hunk 2 above)
Comment (ichard26) [reply src/black/brackets.py:52]: Sounds good to me!

File: src/black/linegen.py (hunk 3 above)
Comment (ichard26) [reply src/black/linegen.py:368]: Yeah, I'd prefer doing this later - tbh I'm a bit tired after all of work! I'll open an issue about this.

Comment (ichard26) [reply src/black/linegen.py:368]: Filed: https://github.com/psf/black/issues/2432.

File: src/black/strings.py (hunk 5 above)
Comment (ichard26) [reply src/black/strings.py:155]: Looking at the source and testing locally, I can confirm that both re and regex cache compiled patterns. Somehow I still observe a 8-9% speed up on that strings list literal file. Running this test script:
```python
import time
from functools import lru_cache

import re
import regex

_cached_regex_compile = lru_cache(regex.compile)

# This is *roughly* how many times regex.compile was called formatting strings-list.py
loops = 410 
pattern = r"\s*\t+\s*(\S)"

re.compile(pattern)
regex.compile(pattern)

t0 = time.perf_counter()
for i in range(1, loops + 1):
    re.compile(pattern)
elapsed = time.perf_counter() - t0
print(f"{loops} cached compiles with re: {elapsed * 1000:.1f}ms")

t0 = time.perf_counter()
for i in range(1, loops + 1):
    regex.compile(pattern)
elapsed = time.perf_counter() - t0
print(f"{loops} cached compiles with regex: {elapsed * 1000:.1f}ms")

t0 = time.perf_counter()
for i in range(1, loops + 1):
    _cached_regex_compile(pattern)
elapsed = time.perf_counter() - t0
print(f"{loops} cached compiles with regex + lru_cache: {elapsed * 1000:.1f}ms")
```
I get:
- 410 cached compiles with re: 0.4ms
- 410 cached compiles with regex: 1.6ms
- 410 cached compiles with regex + lru_cache: 0.1ms

So it seems like this is the source of the slight speedup, but oddly enough this only accounts for ~half of the 8-9% speedup I saw for fmt-strings-list. Maybe the fact this script only repeated compiles one regex is why the delta isn't 1:1. For real life cases this speedup is probably minimal so I wouldn't mind reverting this to be less confusing :)

Comment (JelleZijlstra) [reply src/black/strings.py:155]: Interesting, thanks for the extra information. I looked at `re.compile` and it calls a helper function (`_compile`) and does an `isinstance` call before looking at the cache, and the cache key is a tuple `type(pattern), pattern, flags`. I suppose that adds up to more overhead than whatever lru_cache does internally.

File: src/black/parsing.py (hunk 4 above)
Comment (ichard26) [reply src/black/parsing.py:94]: Ah yea the optimizations I made can be confusing! Profiling found that a lot of time parsing was spent in the Parser.push / Parser.shift methods so optimizing them was critical. Lookups at runtime are expensive and so are calls via the generic Python calling convention so I wanted to avoid them.

File: mypy.ini (hunk 1 above)
Comment (ichard26) [reply mypy.ini:58]: > The behavior will be 100% the same, no need to worry about that.

If this changes I'll be *very* annoyed, but I still trust you so alright.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by ichard26** [comment]
Whoops this isn't PyPy compatible ATM, will fix.
